### DISCUSSION TITLE
actions names in devtools (Redux toolkit)

### DISCUSSION THREAD

**Discussion by Ali-Hussein-dev** [pmndrs/zustand#1668] [Q&A] [answered]
When inspecting Zustand's actions in Redux toolkit, all action names are labeled "anonymous". I tried to use both regular and arrow functions but it didn't change anything. Is there away to change the behaviour?

**Comment by dai-shi** [comment, marked as answer]
You can specify action name in set.

https://github.com/pmndrs/zustand/blob/0b55a3d74b598aba2e41b09854e9237a4ce0f990/readme.md?plain=1#L425

If you don't, it will be anonymous.

https://github.com/pmndrs/zustand/blob/0b55a3d74b598aba2e41b09854e9237a4ce0f990/readme.md?plain=1#L442

**Reply by draeton** [reply]
@dai-shi It doesn't look like the third parameter is included in the type definition. Is that intentional?

**Reply by dai-shi** [reply]
With TypeScript, it will be typed only if `devtools` middleware is used explicitly. Typing is tricky...


https://tsplay.dev/mpK8xN

<img width="581" alt="image" src="https://github.com/pmndrs/zustand/assets/490574/73c77733-0008-4754-ae30-542c5db9d62e">


I think we should update readme.md to follow this pattern to avoid confusion.

**Reply by dai-shi** [reply]
#2132

**Reply by draeton** [reply]
Thanks for the quick response! I was looking for a way to do this with the slice defined in another file, and this is what I came up with after inspecting the devtools types:

_store.ts_

```ts
import { StateCreator } from "zustand";

export type WithDevtools<S> = StateCreator<
    S,
    [["zustand/devtools", never]],
    [],
    S
>;
```

_fooSlice.ts_

```ts
import { WithDevtools } from "./store";

export const createFooSlice: WithDevtools<FooSlice> = (set, get) => ({
```

**Reply by felquis** [reply]
This answer helped me, in combination with the doc https://docs.pmnd.rs/zustand/guides/typescript#create-without-curried-workaround

Possibly a point of improvement is a new section that merges section [slices-pattern](https://docs.pmnd.rs/zustand/guides/typescript#slices-pattern) with [create without curried workaround](https://docs.pmnd.rs/zustand/guides/typescript#create-without-curried-workaround)

Maybe also explaining a little bit of `StateCreator` parameters as well.

Thank you.

**Reply by dai-shi** [reply]
☝️ @dbritto-dev

**Reply by Nimoscmp** [reply]
Yeah, I think this is a good solution as well, you can follow that implementation in the documentation: https://zustand.docs.pmnd.rs/middlewares/devtools#all-action-names-are-labeled-as-'anonymous'

**Comment by gavinthomas-valtech** [comment]
Is there a way to do this with reducers and dispatch pattern? I'm also getting anonymous logged in devtools
I[m using createStore as i need a provider. 
```
return createStore<GlobalStore>()(
    devtools(
      (set) => ({
        ...mergedInitialState,
        dispatch: (event: GlobalStoreEvents) => set((state) => globalStoreReducer(state, event)),
 }),
      { name: 'GlobalStore' },
```

**Reply by dai-shi** [reply]
Have you tried this?
```diff
-         dispatch: (event: GlobalStoreEvents) => set((state) => globalStoreReducer(state, event)),
 }),
+         dispatch: (event: GlobalStoreEvents) => set((state) => globalStoreReducer(state, event), true, 'action name'),
 }),
```

**Reply by gavinthomas-valtech** [reply]
unfortunately i get an error when dispatching tot he global store
TypeError: dispatch is not a function
